lisp/ess-inf.el

```diff
@@ -1134,31 +1134,35 @@ described in `ess-eval-visibly'. STRING need not end with
         (string (ess--run-presend-hooks process string)))
     (inferior-ess--interrupt-subjob-maybe process)
     (inferior-ess-mark-as-busy process)
-    (:override
-     (cond
-      ;; Wait after each line
-      ((eq visibly t)
-       (let ((ess--inhibit-presend-hooks t))
-         (ess-eval-linewise string)))
-      ;; Insert command and eval invisibly
-      ((or (stringp visibly)
-           (eq visibly 'nowait))
-       (with-current-buffer (process-buffer process)
-         (save-excursion
-           (goto-char (process-mark process))
-           (insert-before-markers
-            (propertize (format "%s\n"
-                                (replace-regexp-in-string
-                                 "\n" "\n+ "
-                                 (if (stringp visibly) visibly string)))
-                        'font-lock-face 'comint-highlight-input)))
-         (process-send-string process (ess--concat-new-line-maybe string))))
-      (t
-       (process-send-string process (ess--concat-new-line-maybe string))))
-     (when message
-       (message "%s" message)))))
-
-(ess-defgeneric ess-send-region (process start end &optional visibly message type)
+    (ess-send-string-override process string visibly message type)))
+
+(cl-defgeneric ess-send-string-override (process string visibly message type)
```

Comment:
I think I'm OK with users hooking into the cl-generic machinery if they want. It would be possible to manipulate STRING here for example by defining a `:before` method. I don't have a strong preference though.

Reply:
Ok, but have you thought of a better name than -override? Doesn't feel quite suggestive to me. Ok for internal --override but for public method maybe we could think of something better.

Reply:
I also prefer making as much as possible internal. Users can still hook, they just need to be prepared to change stuff.

By the way, is there a reason that we don't simply use buffer-local `foo-function`s if available? Don't buffer-local hooks give us all the genericity we need?

Reply:
There are at least 4 reasons why cl-generics are better than buffer local hooks:

 - The buffer hooks require setting them in all buffers where you want them. With cl-generics it's nicer because you have to set only ess-dialect which is set anyhow. It saves a lot of hook micromanagement boilerplate code 
- it saves one extra ess-xyz-mode-function for each mode thus not polluting our namespace. cl-generic is one and methods are named the same. 
- The method itself can decide whether to dispatch or not. It potentially can dispatch on much more complex conditions than checking for equality on ess-dialect. With function hooks, the function is just called and that's it. 
- Users could hook on specific methods with :before or others like Alex already pointed out. 

For last two you can probably do hackery with add-function but that seems considerably less elegant than cl-generics. 

> I also prefer making as much as possible internal.

Ok, looks like Alex doesn't have strong preference, so let's make them internal till we have a strong use case for public promotion.

Reply:
That makes good sense, thanks!

Reply:
OK, I've gone through and made them `--override`. I also moved a few to regular defuns that we didn't have any methods for. It's easy enough to move them to generics if we need dialect specific methods in the future. Will merge once the CI passes.